File: min-stack/min-stack-mustafasegf.go
package minstack

type MinStack struct {
	Min   []int
	Stack []int
}

func Constructor() MinStack {
	return MinStack{Stack: make([]int, 0, 30000), Min: make([]int, 0, 30000)}
}

func (this *MinStack) Push(val int) {
	this.Stack = append(this.Stack, val)
	if len(this.Min) == 0 {
		this.Min = append(this.Min, 0)
	} else {
		// last min is less then val
		if this.Stack[this.Min[len(this.Min)-1]] < val {
			this.Min = append(this.Min, this.Min[len(this.Min)-1])
		} else {
			this.Min = append(this.Min, len(this.Stack)-1)
		}
	}
}

func (this *MinStack) Pop() {
	this.Stack = this.Stack[:len(this.Stack)-1]
	this.Min = this.Min[:len(this.Min)-1]
}

func (this *MinStack) Top() int {
	return this.Stack[len(this.Stack)-1]
}

func (this *MinStack) GetMin() int {
	return this.Stack[this.Min[len(this.Min)-1]]
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(val);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */

File: min-stack/min-stack-mustafasegf.rs
use std::cmp::{min, max};

struct MinStack {
  stack : Vec<(i32, i32)>
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MinStack {

    fn new() -> Self {
      Self {
        stack: Vec::with_capacity(30_000)
      }
    }
    
    fn push(&mut self, val: i32) {
        match self.stack.is_empty() {
          true => self.stack.push((val, val)),
          false => {
            let min_val = min(val, self.get_min());
            self.stack.push((val, min_val));
          }
        }
    }
    
    fn pop(&mut self) {
      self.stack.pop();
    }
    
    fn top(&self) -> i32 {
      return match self.stack.last() {
        Some(val) => val.0,
        _ => 0,
      };
    }
    
    fn get_min(&self) -> i32 {
      return match self.stack.last() {
        Some(val) => val.1,
        _ => 0,
      };
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * let obj = MinStack::new();
 * obj.push(val);
 * obj.pop();
 * let ret_3: i32 = obj.top();
 * let ret_4: i32 = obj.get_min();
 */
